From d4229a3ab952ea2efa074703c881de0df6b921d9 Mon Sep 17 00:00:00 2001
From: Marcelo Leitner <mleitner@redhat.com>
Date: Tue, 21 Oct 2014 16:20:30 -0400
Subject: [net] netfilter: nf_tables: add devgroup support in meta expresion

Message-id: <20141021162030.22888.39518.stgit@localhost.localdomain>
Patchwork-id: 98000
O-Subject: [RHEL7 net PATCH 103/124] netfilter: nf_tables: add devgroup support in meta expresion
Bugzilla: 1148041 1155088
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

Backport of upstream commit: 3045d76070abe725dbb7fd8ff39c27b820d5a7eb

 commit 3045d76070abe725dbb7fd8ff39c27b820d5a7eb
 Author: Ana Rey <anarey@gmail.com>
 Date:   Tue Sep 2 20:36:14 2014 +0200

    netfilter: nf_tables: add devgroup support in meta expresion

    Add devgroup support to let us match device group of a packets incoming
    or outgoing interface.

    Signed-off-by: Ana Rey <anarey@gmail.com>
    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Marcelo Ricardo Leitner <mleitner@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/include/uapi/linux/netfilter/nf_tables.h b/include/uapi/linux/netfilter/nf_tables.h
index c9b6f00..c000947 100644
--- a/include/uapi/linux/netfilter/nf_tables.h
+++ b/include/uapi/linux/netfilter/nf_tables.h
@@ -573,6 +573,8 @@ enum nft_exthdr_attributes {
  * @NFT_META_BRI_OIFNAME: packet output bridge interface name
  * @NFT_META_PKTTYPE: packet type (skb->pkt_type), special handling for loopback
  * @NFT_META_CPU: cpu id through smp_processor_id()
+ * @NFT_META_IIFGROUP: packet input interface group
+ * @NFT_META_OIFGROUP: packet output interface group
  */
 enum nft_meta_keys {
  NFT_META_LEN,
@@ -596,6 +598,8 @@ enum nft_meta_keys {
  NFT_META_BRI_OIFNAME,
  NFT_META_PKTTYPE,
  NFT_META_CPU,
+ NFT_META_IIFGROUP,
+ NFT_META_OIFGROUP,
 };
 
 /**
diff --git a/net/netfilter/nft_meta.c b/net/netfilter/nft_meta.c
index 843e099..1e7c076 100644
--- a/net/netfilter/nft_meta.c
+++ b/net/netfilter/nft_meta.c
@@ -155,6 +155,16 @@ void nft_meta_get_eval(const struct nft_expr *expr,
  case NFT_META_CPU:
   dest->data[0] = smp_processor_id();
   break;
+ case NFT_META_IIFGROUP:
+  if (in == NULL)
+   goto err;
+  dest->data[0] = in->group;
+  break;
+ case NFT_META_OIFGROUP:
+  if (out == NULL)
+   goto err;
+  dest->data[0] = out->group;
+  break;
  default:
   WARN_ON(1);
   goto err;
@@ -228,6 +238,8 @@ int nft_meta_get_init(const struct nft_ctx *ctx,
 #endif
  case NFT_META_PKTTYPE:
  case NFT_META_CPU:
+ case NFT_META_IIFGROUP:
+ case NFT_META_OIFGROUP:
   break;
  default:
   return -EOPNOTSUPP;
-- 
1.7.1